### PR TITLE
[Dragon Hunt] Almost kill ExtruderManager

### PR DESCRIPTION
**Pull request by LipuFei**
Followup of https://github.com/Ultimaker/Cura/pull/4431

Removed most functions in ExtruderManager. Unfortunately, many of them have to be placed into `MachineManager`.

Major difference: Now `MachineManager` maintains the active extruder.

### CODE REVIEW

**Comment by Appesteijn** [comment]
CURA-5782

**Comment by LipuFei** [comment]
@nallath 
 - Updated this branch based on master
 - Fixed some problems caused by some refactors in the previous PR
 - Resolved merge conflicts

**Comment by Ghostkeeper** [comment]
Next time, it would help to move such functions one by one in separate commits. If something goes wrong, we can track much easier where the bug came from.

File: cura/Settings/MachineManager.py [hunk 1]
```diff
@@ -729,22 +729,41 @@ def renameMachine(self, machine_id: str, new_name: str) -> None:
...
-        ExtruderManager.getInstance().removeMachineExtruders(machine_id)
-        containers = CuraContainerRegistry.getInstance().findInstanceContainersMetadata(type = "user", machine = machine_id)
+
+        for extruder_stack in extruder_stack_list:
+            container_registry.removeContainer(extruder_stack.userChanges.getId())
+            container_registry.removeContainer(extruder_stack.definitionChanges.getId())
+            container_registry.removeContainer(extruder_stack.getId())
+
+        container_registry.removeContainer(global_stack.definitionChanges.getId())
+        containers = container_registry.findInstanceContainersMetadata(type = "user", machine = machine_id)
```
Comment (Ghostkeeper) [inline cura/Settings/MachineManager.py]: Can't you just remove `global_stack.userChanges.getId()` from the registry? Why this db query?

File: cura/Settings/SettingInheritanceManager.py [hunk 2]
```diff
@@ -1,40 +1,41 @@
...
 class SettingInheritanceManager(QObject):
     def __init__(self, parent = None) -> None:
         super().__init__(parent)
-        Application.getInstance().globalContainerStackChanged.connect(self._onGlobalContainerChanged)
-        self._global_container_stack = None  # type: Optional[ContainerStack]
-        self._settings_with_inheritance_warning = []  # type: List[str]
-        self._active_container_stack = None  # type: Optional[ExtruderStack]
+
+        self._application = Application.getInstance()
+        self._machine_manager = self._application.getMachineManager()
```
Comment (Ghostkeeper) [inline cura/Settings/SettingInheritanceManager.py]: Why are we caching these? There are now two sources of truth.

File: cura/Settings/SettingOverrideDecorator.py [hunk 3]
```diff
@@ -34,21 +38,26 @@ class SettingOverrideDecorator(SceneNodeDecorator):
 
     def __init__(self):
         super().__init__()
+
+        self._application = Application.getInstance()
+        self._machine_manager = self._application.getMachineManager()
```
Comment (Ghostkeeper) [inline cura/Settings/SettingOverrideDecorator.py]: Why are we caching these? There are now two sources of truth.

File: plugins/PostProcessingPlugin/scripts/Stretch.py [hunk 4]
```diff
@@ -452,6 +452,8 @@ class Stretch(Script):
     def __init__(self):
         super().__init__()
 
+        self._application = Application.getInstance()
```
Comment (Ghostkeeper) [inline plugins/PostProcessingPlugin/scripts/Stretch.py]: Why do we need to cache this? `Application.getInstance()` is just as nice, and there will only be a single source of truth.

File: plugins/SimulationView/SimulationPass.py [hunk 5]
```diff
@@ -37,7 +34,8 @@ def __init__(self, width, height):
         self._switching_layers = True # It tracks when the user is moving the layers' slider
         self._gl = OpenGL.getInstance().getBindingsObject()
         self._scene = Application.getInstance().getController().getScene()
-        self._extruder_manager = ExtruderManager.getInstance()
+
+        self._application = Application.getInstance()
```
Comment (Ghostkeeper) [inline plugins/SimulationView/SimulationPass.py]: Why are we caching this? I'd prefer to have only a single source of truth.

File: cura/Settings/MachineManager.py [hunk 6]
```diff
@@ -535,6 +517,21 @@ def activeMachineNetworkGroupName(self) -> str:
     def activeMachine(self) -> Optional["GlobalStack"]:
         return self._global_container_stack
 
+    @pyqtSlot(int)
+    def setActiveExtruderPosition(self, position: int) -> None:
+        if self._active_container_stack is None or self._active_container_stack.getMetaDataEntry("position") != position:
+            self._active_container_stack = self._global_container_stack.extruders[str(position)]
+            self.activeStackChanged.emit()
+
+    def getActiveExtruderPosition(self) -> int:
```
Comment (Ghostkeeper) [inline cura/Settings/MachineManager.py]: Needs to be called an `Index` according to the code style.

File: cura/Settings/MachineManager.py [hunk 7]
```diff
@@ -535,6 +517,21 @@ def activeMachineNetworkGroupName(self) -> str:
     def activeMachine(self) -> Optional["GlobalStack"]:
         return self._global_container_stack
 
+    @pyqtSlot(int)
+    def setActiveExtruderPosition(self, position: int) -> None:
```
Comment (Ghostkeeper) [inline cura/Settings/MachineManager.py]: Needs to be called an `Index` according to the code style.

File: plugins/CuraEngineBackend/StartSliceJob.py [hunk 8]
```diff
@@ -453,11 +457,11 @@ def _handlePerObjectSettings(self, node: CuraSceneNode, message: Arcus.PythonMes
         for key in changed_setting_keys:
             setting = message.addRepeatedMessage("settings")
             setting.name = key
-            extruder = int(round(float(stack.getProperty(key, "limit_to_extruder"))))
+            extruder_position = int(round(float(stack.getProperty(key, "limit_to_extruder"))))
```
Comment (Ghostkeeper) [inline plugins/CuraEngineBackend/StartSliceJob.py]: _index

File: plugins/PerObjectSettingsTool/PerObjectSettingVisibilityHandler.py [hunk 9]
```diff
@@ -65,27 +65,25 @@ def setVisible(self, visible):
...
+        global_stack = machine_manager.activeMachine
+
         # Add all instances that are not added, but are in visibility list
         for item in visible:
             if not settings.getInstance(item): # Setting was not added already.
                 definition = self._stack.getSettingDefinition(item)
                 if definition:
                     new_instance = SettingInstance(definition, settings)
-                    stack_nr = -1
+                    extruder_position = machine_manager.defaultExtruderPosition
```
Comment (Ghostkeeper) [inline plugins/PerObjectSettingsTool/PerObjectSettingVisibilityHandler.py]: _index

File: cura/Settings/MachineManager.py [hunk 10]
```diff
@@ -729,22 +729,41 @@ def renameMachine(self, machine_id: str, new_name: str) -> None:
     @pyqtSlot(str)
     def removeMachine(self, machine_id: str) -> None:
```
Comment (Ghostkeeper) [inline cura/Settings/MachineManager.py]: You could test this.

File: cura/Settings/MachineManager.py (hunk 7 above)
Comment (Ghostkeeper) [inline cura/Settings/MachineManager.py]: You could test this.

File: cura/Settings/GlobalStack.py [hunk 11]
```diff
@@ -202,6 +203,22 @@ def getHasVariants(self) -> bool:
...
+    # machine.
+    def getMachineExtruderStacks(self) -> List["ExtruderStack"]:
+        machine_extruder_count = self.getProperty("machine_extruder_count", "value")
+        result_list = []
+        for extruder_position in range(machine_extruder_count):
+            result_list.append(self.extruders[str(extruder_position)])
+        return result_list
+
+    # Gets values of a setting for all extruders.
+    def getValuesInAllExtruders(self, setting_key: str, property_name: str) -> List[Any]:
```
Comment (Ghostkeeper) [inline cura/Settings/GlobalStack.py]: You could test this.

File: cura/Settings/GlobalStack.py [hunk 12]
```diff
@@ -202,6 +203,22 @@ def getHasVariants(self) -> bool:
     def getHasMachineQuality(self) -> bool:
         return parseBool(self.getMetaDataEntry("has_machine_quality", False))
 
+    # Returns the ExtruderStacks that can be used on this machine, i.e. the number of extruders specified for this
+    # machine.
+    def getMachineExtruderStacks(self) -> List["ExtruderStack"]:
```
Comment (Ghostkeeper) [inline cura/Settings/GlobalStack.py]: You could test this.

**Pull request closed without merging**